File: python/875_Koko_Eating_Bananas.py
from math import ceil

class Koko_Eating_Bananas:
  def minEatingSpeed(self, piles: list[int], h: int) -> int:
    low = 1
    high = max(piles)
    med = int((low + high) / 2)
    while (low < high):
      numHours = 0
      for pile in piles:
        numHours += ceil(pile / med)
      if numHours <= h:
        high = med
      else:
        low = med + 1
      med = int((low + high) / 2)
    return med

File: c++/875_Koko_Eating_Bananas.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Koko_Eating_Bananas {
public:
  int minEatingSpeed(vector<int>& piles, int h) {
    int low = 1;
    int high = *max_element(piles.begin(), piles.end());
    int mid = (low + high) / 2;

    while (low < high) {
      int numHours = 0;
      for (int pile : piles) {
        numHours += (pile / mid);
        if (pile % mid != 0) {
          numHours ++;
        }
      }
      if (numHours <= h) {
        high = mid;
      } else {
        low = mid + 1;
      }
      mid = (low + high) / 2;
    }
    return mid;
  }
};